service/frontend/adminHandler.go

```diff
@@ -1928,19 +1929,36 @@ func (adh *AdminHandler) getWorkflowCompletionEvent(
 		return nil, err
 	}
 
-	resp, err := adh.persistenceExecutionManager.ReadHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
-		ShardID:     shardID,
-		BranchToken: currentVersionHistory.GetBranchToken(),
-		MinEventID:  executionInfo.CompletionEventBatchId,
-		MaxEventID:  completionEventID + 1,
-		PageSize:    1,
-	})
-	if err != nil {
-		return nil, err
+	var historyEvents []*historypb.HistoryEvent
+	if adh.config.readEventsFromHistory(adh.metricsHandler) {
+		resp, err := adh.historyClient.ReadHistoryBranch(ctx, &historyservice.ReadHistoryBranchRequest{
+			NamespaceId: namespaceID.String(),
+			ShardId:     shardID,
+			BranchToken: currentVersionHistory.GetBranchToken(),
+			MinEventId:  executionInfo.CompletionEventBatchId,
+			MaxEventId:  completionEventID + 1,
+			PageSize:    1,
+		})
+		if err != nil {
+			return nil, err
+		}
+		historyEvents = resp.HistoryEvents
+	} else {
+		resp, err := adh.persistenceExecutionManager.ReadHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
+			ShardID:     shardID,
+			BranchToken: currentVersionHistory.GetBranchToken(),
+			MinEventID:  executionInfo.CompletionEventBatchId,
+			MaxEventID:  completionEventID + 1,
+			PageSize:    1,
+		})
+		if err != nil {
+			return nil, err
+		}
+		historyEvents = resp.HistoryEvents
 	}
```

Comment:
nit: extract to an fn to avoid `var historyEvents []*historypb.HistoryEvent` declaration